clients/drcachesim/scheduler/scheduler.h

```diff
@@ -0,0 +1,562 @@
+                         std::vector<range_t> regions_of_interest)
+            : struct_size(sizeof(input_workload_t))
+            , path(trace_path)
+        {
+            thread_modifiers.push_back(input_thread_info_t(regions_of_interest));
+        }
+        /** Size of the struct for binary-compatible additions. */
+        size_t struct_size = sizeof(input_workload_t);
+        /**
+         * A directory of trace files or a single trace file.
+         * A reader will be opened for each input file and its init() function
+         * will be called; that function will not block (variants where it
+         * does block, such as for IPC, should leave 'path' empty and use 'reader'
+         * below instead).
+         */
+        std::string path;
+        /**
+         * An alternative to passing in a path and having the scheduler open the file(s)
+         * there is to directly pass in a reader, reader_end, and a thread id for that
+         * reader.  These are only considered if 'path' is empty.  The scheduler will
+         * call the init() function for 'reader' at the time of the first access to
+         * it through an output stream (supporting IPC readers whose init() blocks).
+         */
+        std::unique_ptr<ReaderType> reader;
+        /** The end reader for 'reader'. */
+        std::unique_ptr<ReaderType> reader_end;
+        /** The thread identifier for 'reader'. */
+        memref_tid_t reader_tid = INVALID_THREAD_ID;
+
+        /** Scheduling modifiers for the threads in this workload. */
+        std::vector<input_thread_info_t> thread_modifiers;
+    };
+
+    /** Controls the primary mode of the scheduler regarding the input streams. */
+    enum stream_type_t {
+        /**
+         * Each input stream is mapped without interleaving directly onto a single
+         * output stream (i.e., no input will appear on more than one output),
+         * supporting parallel analysis.  The only strategy supported here is
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::SCHEDULE_RUN_TO_COMPLETION.
+         */
+        STREAM_BY_INPUT_SHARD,
+        /**
+         * Each input stream is assumed to contain markers indicating how it was
+         * originally scheduled.  This recorded schedule is replayed.  This requires an
+         * output stream count equal to the number of cores occupied by the input stream
+         * set.  The only strategy supported here is
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::SCHEDULE_INTERLEAVE_AS_RECORDED.
+         */
+        STREAM_BY_RECORDED_CPU,
+        /**
+         * The input streams are scheduling using a new synthetic schedule onto the
+         * output streams.  Any input markers indicating how the software threads were
+         * originally scheduled during tracing are ignored.
+         */
+        STREAM_BY_SYNTHETIC_CPU,
+    };
+
+    /** Controls the output of the scheduler. */
+    enum schedule_strategy_t {
```

Comment:
stream_type_t and schedule_strategy_t are so tightly coupled that I wonder if it really makes sense for them be separate.  It seems every schedule_strategy only works with a single stream_type.

Reply:
This is one thing I wanted to discuss offline as well, whether this separation makes sense.

Reply:
Per offline discussion, we replaced both of these with the mapping and dependencies:

Inter-input dependency enum
 + None => pure parallel
 + Timestamps => replay as recorded
 + Inferred data deps

Input-to-output mapping
 + NO_MIGRATION (N-to-1: any one input is only ever on one output) instead of “run to completion w/ no interleaving”
 + As recorded
 + Synthetic